File: LeetCodeAlgos/single-number.rs
// Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

// You must implement a solution with a linear runtime complexity and use only constant extra space.


struct Solution {}
impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        let mut ans = 0;
        for num in nums{
            ans ^= num;
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_number() {
        assert_eq!(Solution::single_number(vec![2, 2, 1]), 1);
    }

    #[test]
    fn test_single_number2() {
        assert_eq!(Solution::single_number(vec![4, 1, 2, 1, 2]), 4);
    }

    #[test]
    fn test_single_number3() {
        assert_eq!(Solution::single_number(vec![1]), 1);
    }
}

File: LeetCodeAlgos/Java/singleNumber.java
class Solution {
    public int singleNumber(int[] nums) {
        int ans = 0;
      for (int num: nums){
        ans ^= num;
      }
      return ans;
    }
}

class singleNumber {
  public static void main(String[] args) {
    Solution s = new Solution();
    System.out.println(s.singleNumber(new int[]{1,2,2}));
    System.out.println(s.singleNumber(new int[]{4,1,2,1,2}));
    System.out.println(s.singleNumber(new int[]{1}));
  }
}